src/server/scripts/EasternKingdoms/zone_tol_barad.cpp

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) 2008-2016 TrinityCore <http://www.trinitycore.org/>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "BattlefieldMgr.h"
+#include "BattlefieldTB.h"
+#include "Battlefield.h"
+#include "ScriptSystem.h"
+#include "WorldSession.h"
+#include "ScriptedCreature.h"
+#include "ScriptedGossip.h"
+#include "SpellScript.h"
+#include "Player.h"
+
+enum TBSpiritGuide
+{
+    GOSSIP_OPTION_ID_SLAGWORKS = 0,
+    GOSSIP_OPTION_ID_IRONCLAD_GARRISON = 1,
+    GOSSIP_OPTION_ID_WARDENS_VIGIL = 2,
+    GOSSIP_OPTION_ID_EAST_SPIRE = 3,
+    GOSSIP_OPTION_ID_WEST_SPIRE = 4,
+    GOSSIP_OPTION_ID_SOUTH_SPIRE = 5,
+
+    SPELL_CHANNEL_SPIRIT_HEAL = 22011,
+};
+
+class npc_tb_spirit_guide : public CreatureScript
+{
+    public:
+        npc_tb_spirit_guide() : CreatureScript("npc_tb_spirit_guide") { }
+
+        bool OnGossipHello(Player* player, Creature* creature) override
+        {
+            Battlefield* tb = sBattlefieldMgr->GetBattlefieldByBattleId(BATTLEFIELD_BATTLEID_TB);
+            if (!tb || !tb->IsWarTime())
+                return true;
+
+            if (creature->IsQuestGiver())
+                player->PrepareQuestMenu(creature->GetGUID());
+
+            uint32 entry = creature->GetEntry();
+            if (entry != NPC_TB_GY_SPIRIT_SLAGWORKS_A && entry != NPC_TB_GY_SPIRIT_SLAGWORKS_H)
```

Comment:
Not sure, but if there is a dedicated spell to teleport the player to WorldSafeLocs, it can be done in SAI fairly easy.

Other than that, I think Aok refers to what joschiwald usually asks me to do, which is to have all of `OnGossipHello` handled in the DB and use an adapted form of
```cpp
void sGossipSelect(Player* player, uint32 menuId, uint32 gossipListId) override
```
instead of the `OnGossipSelect()` hook below.